pkg/client/patch.go

```diff
@@ -96,8 +96,42 @@ func (s *mergeFromPatch) Type() types.PatchType {
 	return types.MergePatchType
 }
 
+func (*mergeFromPatch) data(original, modified Object, opts MergeFromOptions) ([]byte, error) {
+	if opts.OptimisticLock {
+		version := original.GetResourceVersion()
+		if len(version) == 0 {
+			return nil, fmt.Errorf("cannot use OptimisticLock, object %q does not have any resource version we can use", original)
+		}
+
+		original = original.DeepCopyObject().(Object)
+		original.SetResourceVersion("")
+
+		modified = modified.DeepCopyObject().(Object)
+		modified.SetResourceVersion(version)
+	}
+
+	originalJSON, err := json.Marshal(original)
+	if err != nil {
+		return nil, err
+	}
+
+	modifiedJSON, err := json.Marshal(modified)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonpatch.CreateMergePatch(originalJSON, modifiedJSON)
+}
+
 // Data implements Patch.
 func (s *mergeFromPatch) Data(obj runtime.Object) ([]byte, error) {
```

Comment:
Should we just move this to use `client.Object` instead? I don't think we necessarily want to support all runtime objects, because we still require ObjectMeta.

Reply:
cc @alvaroaleman

Reply:
Yeah, sounds like a good idea

Reply:
Done in aef159a.